### PR TITLE
Add type hints to `test_formparsers.py`

### PR DESCRIPTION
**Pull request by TechNiick**
# Summary

Related to [this](https://github.com/encode/starlette/pull/2445)
Type annotation added to `test_formparsers.py`

# Checklist

- [ ] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [ ] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

File: tests/test_formparsers.py [hunk 1]
```diff
@@ -304,7 +328,10 @@ def test_multipart_request_mixed_files_and_data(tmpdir, test_client_factory):
     }
 
 
-def test_multipart_request_with_charset_for_filename(tmpdir, test_client_factory):
+def test_multipart_request_with_charset_for_filename(
+    tmpdir: Path,
+    test_client_factory: typing.Any,
```
Comment (Kludex) [inline tests/test_formparsers.py]: Why Any on those?

**Review by Kludex — CHANGES_REQUESTED** [review]

File: tests/test_formparsers.py [hunk 2]
```diff
@@ -574,7 +643,11 @@ def test_too_many_files_single_field_raise(app, expectation, test_client_factory
         (Starlette(routes=[Mount("/", app=app)]), does_not_raise()),
     ],
 )
-def test_too_many_files_and_fields_raise(app, expectation, test_client_factory):
+def test_too_many_files_and_fields_raise(
+    app: ASGIApp,
+    expectation: typing.ContextManager[TestCase],
```
Comment (Kludex) [inline tests/test_formparsers.py]: I don't think `TestCase` is the right type here... 👀

**Review by Kludex — CHANGES_REQUESTED** [review]

File: tests/test_formparsers.py (hunk 1 above)
Comment (TechNiick) [reply tests/test_formparsers.py:333]: Adding `TestClientFactory` on those results in this error: 
`Argument "data" to "post" of "TestClient" has incompatible type "bytes"; expected "Optional[Mapping[str, Union[str, Iterable[str]]]]"`

File: tests/test_formparsers.py [hunk 3]
```diff
@@ -122,13 +127,19 @@ async def app(scope, receive, send):
     return app
 
 
-def test_multipart_request_data(tmpdir, test_client_factory):
+def test_multipart_request_data(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 4]
```diff
@@ -122,13 +127,19 @@ async def app(scope, receive, send):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post("/", data={"some": "data"}, files=FORCE_MULTIPART)
     assert response.json() == {"some": "data"}
 
 
-def test_multipart_request_files(tmpdir, test_client_factory):
+def test_multipart_request_files(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 5]
```diff
@@ -146,7 +157,10 @@ def test_multipart_request_files(tmpdir, test_client_factory):
         }
 
 
-def test_multipart_request_files_with_content_type(tmpdir, test_client_factory):
+def test_multipart_request_files_with_content_type(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 6]
```diff
@@ -164,7 +178,10 @@ def test_multipart_request_files_with_content_type(tmpdir, test_client_factory):
         }
 
 
-def test_multipart_request_multiple_files(tmpdir, test_client_factory):
+def test_multipart_request_multiple_files(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 7]
```diff
@@ -194,7 +211,10 @@ def test_multipart_request_multiple_files(tmpdir, test_client_factory):
         }
 
 
-def test_multipart_request_multiple_files_with_headers(tmpdir, test_client_factory):
+def test_multipart_request_multiple_files_with_headers(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 8]
```diff
@@ -266,13 +286,16 @@ def test_multi_items(tmpdir, test_client_factory):
         }
 
 
-def test_multipart_request_mixed_files_and_data(tmpdir, test_client_factory):
+def test_multipart_request_mixed_files_and_data(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 9]
```diff
@@ -266,13 +286,16 @@ def test_multi_items(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post(
         "/",
         data=(
             # data
-            b"--a7f7ac8d4e2e437c877bb7b8d7cc549c\r\n"
+            b"--a7f7ac8d4e2e437c877bb7b8d7cc549c\r\n"  # type: ignore
```
Comment (Kludex) [inline tests/test_formparsers.py:292]: Why the ignore here?

File: tests/test_formparsers.py [hunk 10]
```diff
@@ -304,13 +327,16 @@ def test_multipart_request_mixed_files_and_data(tmpdir, test_client_factory):
     }
 
 
-def test_multipart_request_with_charset_for_filename(tmpdir, test_client_factory):
+def test_multipart_request_with_charset_for_filename(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 11]
```diff
@@ -333,13 +359,16 @@ def test_multipart_request_with_charset_for_filename(tmpdir, test_client_factory
     }
 
 
-def test_multipart_request_without_charset_for_filename(tmpdir, test_client_factory):
+def test_multipart_request_without_charset_for_filename(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 12]
```diff
@@ -361,12 +390,15 @@ def test_multipart_request_without_charset_for_filename(tmpdir, test_client_fact
     }
 
 
-def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
+def test_multipart_request_with_encoded_value(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 13]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
     assert response.json() == {"value": "Transférer"}
 
 
-def test_urlencoded_request_data(tmpdir, test_client_factory):
+def test_urlencoded_request_data(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 14]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post("/", data={"some": "data"})
     assert response.json() == {"some": "data"}
 
 
-def test_no_request_data(tmpdir, test_client_factory):
+def test_no_request_data(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 15]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post("/")
     assert response.json() == {}
 
 
-def test_urlencoded_percent_encoding(tmpdir, test_client_factory):
+def test_urlencoded_percent_encoding(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 16]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post("/", data={"some": "da ta"})
     assert response.json() == {"some": "da ta"}
 
 
-def test_urlencoded_percent_encoding_keys(tmpdir, test_client_factory):
+def test_urlencoded_percent_encoding_keys(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 17]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app)
     response = client.post("/", data={"so me": "data"})
     assert response.json() == {"so me": "data"}
 
 
-def test_urlencoded_multi_field_app_reads_body(tmpdir, test_client_factory):
+def test_urlencoded_multi_field_app_reads_body(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

File: tests/test_formparsers.py [hunk 18]
```diff
@@ -382,50 +414,68 @@ def test_multipart_request_with_encoded_value(tmpdir, test_client_factory):
...
+) -> None:
     client = test_client_factory(app_read_body)
     response = client.post("/", data={"some": "data", "second": "key pair"})
     assert response.json() == {"some": "data", "second": "key pair"}
 
 
-def test_multipart_multi_field_app_reads_body(tmpdir, test_client_factory):
+def test_multipart_multi_field_app_reads_body(
+    tmpdir: Path,
+    test_client_factory: TestClientFactory,
```
Comment (Kludex) [inline tests/test_formparsers.py]:
```suggestion
    tmpdir: Path, test_client_factory: TestClientFactory
```

**Review by Kludex — CHANGES_REQUESTED** [review]

**Review by Kludex — APPROVED** [review]

**Pull request merged**